core/src/main/java/com/amazon/opendistroforelasticsearch/sql/expression/function/FunctionDSL.java

```diff
@@ -0,0 +1,160 @@
+ *    A copy of the License is located at
+ *
+ *        http://www.apache.org/licenses/LICENSE-2.0
+ *
+ *    or in the "license" file accompanying this file. This file is distributed
+ *    on an "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either
+ *    express or implied. See the License for the specific language governing
+ *    permissions and limitations under the License.
+ *
+ */
+
+package com.amazon.opendistroforelasticsearch.sql.expression.function;
+
+import com.amazon.opendistroforelasticsearch.sql.data.model.ExprValue;
+import com.amazon.opendistroforelasticsearch.sql.data.model.ExprValueUtils;
+import com.amazon.opendistroforelasticsearch.sql.data.type.ExprType;
+import com.amazon.opendistroforelasticsearch.sql.expression.Expression;
+import com.amazon.opendistroforelasticsearch.sql.expression.FunctionExpression;
+import com.amazon.opendistroforelasticsearch.sql.expression.env.Environment;
+import java.util.Arrays;
+import java.util.Collections;
+import java.util.List;
+import java.util.function.Function;
+import java.util.stream.Collectors;
+import lombok.experimental.UtilityClass;
+import org.apache.commons.lang3.tuple.Pair;
+
+/**
+ * Function Define Utility.
+ */
+@UtilityClass
+public class FunctionDSL {
+  /**
+   * Define overloaded function with implementation.
+   *
+   * @param functionName function name.
+   * @param functions    a list of function implementation.
+   * @return FunctionResolver.
+   */
+  public FunctionResolver define(FunctionName functionName,
+                                 Function<FunctionName, Pair<FunctionSignature,
+                                     FunctionBuilder>>... functions) {
+    return define(functionName, Arrays.asList(functions));
+  }
+
+  /**
+   * Define overloaded function with implementation.
+   *
+   * @param functionName function name.
+   * @param functions    a list of function implementation.
+   * @return FunctionResolver.
+   */
+  public FunctionResolver define(FunctionName functionName,
+                                 List<Function<FunctionName, Pair<FunctionSignature,
+                                     FunctionBuilder>>> functions) {
+
+    FunctionResolver.FunctionResolverBuilder builder = FunctionResolver.builder();
+    builder.functionName(functionName);
+    for (Function<FunctionName, Pair<FunctionSignature, FunctionBuilder>> func : functions) {
+      Pair<FunctionSignature, FunctionBuilder> functionBuilder = func.apply(functionName);
+      builder.functionBundle(functionBuilder.getKey(), functionBuilder.getValue());
+    }
+    return builder.build();
+  }
+
+  /**
+   * Unary Function Implementation.
+   *
+   * @param function   {@link ExprValue} based unary function.
+   * @param returnType return type.
+   * @param argsType   argument type.
+   * @return Unary Function Implementation.
+   */
+  public SerializableFunction<FunctionName, Pair<FunctionSignature, FunctionBuilder>> impl(
```

Comment:
np: the `impl()` for binary function returns `Function` as below rather than `SerializableFunction`?

Reply:
Done.